src/shared.py

```diff
@@ -0,0 +1,30 @@
+import logging
+import time
+from functools import wraps
+from typing import Callable, Tuple, Type
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+def exponential_backoff(
+    attempts: int, initial_delay: int, factor: int, exception_types: Tuple[Type[Exception], ...]
+) -> Callable:
+    def exponential_backoff_decorator(func: Callable) -> Callable:

```

Comment:
This is great. Decorators can feel complicated sometimes when you open them up - so I'd just recommend a docstring and a couple of inline comments to to explain why it's doing what it does.

Not too much, just a help for future you.